### ISSUE TITLE
Allow a ring that spans a network over NAT

### ISSUE THREAD

**Issue by krnowak** [habitat-sh/habitat#4211]
- network abstraction
  - [x] replace direct use of UDP and zeromq sockets and socket addresses with abstract SWIM and gossip message senders and receivers
  - [x] implement an abstraction for real-world usage (with UDP and zeromq sockets using SocketAddr as addresses)
  - [x] use the real-world abstraction in supervisor and tests
- test infrastructure for spanning ring
  - [x] channel-based network abstraction for testing purposes
  - [x] NAT simulation
    - set up simulated private networks, expose addresses outside the private network, start Butterfly servers with a given address, route/block the messages sent between the servers
  - [ ] generating temporary addresses on the NAT boundary
    - mostly for testing if supervisors don't try to talk other supervisors without being exposed
  - [ ] customizing NAT features (ban hairpinning, port- or address-mapping stability)
    - rather low-priority, to test various types of NATs
- spanning ring
  - [x] zone concept
    - it's a supervisor's view of the private networks hierarchy
  - [x] send zone rumors
  - [x] zone merging, zone negotiation within a single private network, basic zone negotiation across private networks, tests
    - the three items are lumped into one as they all are quite related and touch the same code
    - tests are done
    - got the first two items to work, working on the third broke the first two
    - largest item in the work
  - [ ] add an "indirect" health status
    - can be done a bit later, not needed for the demo
  - [ ] zone negotiation across private networks with hidden supervisors
    - can be done a bit later, not needed for the demo
  - [ ] set up additional addresses and ports
    - for butterfly servers:
      - these are addresses of a supervisor that is exposed
      - the address is optional, it can be discovered, ports are not - they need to be specified
      - directly from command line
      - from a watched file
    - for services
      - a bit similar to additional addresses for butterfly servers
      - each port needs to be named like the variable in the exposed configuration it relates to
    - IN PROGRESS
  - [ ] update configuration template helpers
    - likely sys.ip should return an IP that the service in a specific zone can actually use to talk to the other service.
  - [ ] add sibling zones
    - to allow a scenario where we have two k8s clusters, each are running some services managed by supervisors and those are exposed and then bound to each other
  - [ ] add validation for incoming zone and member information
    - make sure that UUIDs sent as strings are actually valid
  - [ ] handle the case when the maintainer of the current zone disappears

Old old stuff below:

These notes are still work in progress. I hope to get to some concrete steps to be done soon.

Currently it is not possible to have a ring that contains supervisors talking to each other over NAT.

- supervisor behind NAT needs to know which address it got on the NAT boundary
  - so it can send a proper information in the Ping message.

- questions:
  - do we want the supervisors behind the NAT to only use the exposed IP for communication?
    - if yes, that would assume that all the supervisors behind NAT need to be exposed
      - also, if we have NAT behind another NAT, then all the supervisors needs to be exposed over the uppermost NAT and use that IP in the ring communication.
    - if not, we need a way of partitioning the ring into zones (where each zone only contains supervisors that can talk directly in both ways) and setting up some supervisors in the zone to act as a forwarder when supervisors are communicating over the NAT boundary.
      - this is quite involved and I was spending my time mostly on this scenario, as exposing all the supervisors may not scale.
      - this will certainly mean some changes in the swim/gossip messages, hopefully backward compatible.
      - what may not be possible is to extend an already existing ring with older supervisors so it starts spanning the NAT boundary.
        - might be a bit tricky to detect that…

**Comment by krnowak** [comment]
Related issue: #2420

**Comment by tashimi** [comment]
This is an interesting use case, pinging @fnichol @christophermaier @reset for their thoughts on how to achieve this. The idea is how to handle port management from the kubernetes cluster to the outside world - it would be easier to have 1 port to the outside world, instead of needing to have an address for every supervisor within the kubernetes cluster.

**Comment by krnowak** [comment]
Sorry for the wall of text.

Currently ring is defined basically as a list of members. A member is
described with an ID, address and ports (1) and other data. If a
supervisor receives a message with that list, it knows how to talk to
other supervisors in the ring. A member contains information about a
supervisor, so they are used interchangeably in the following text.

Currently I'd like to add another thing that will be a part of a ring
definition - zones. Zone is a part of a ring. It does not overlap with
other zones. In other words - zone can contain many supervisors, but
one supervisor belongs to only one zone. Zone is meant to describe the
network of supervisors that can easily talk to each other, that is -
all the supervisor in a zone needs to talk to other supervisor in the
same zone is just its IP address and port. The communication between
zones might be a bit more involved, which is true when we want to
talk with a service behind a NAT. An example of it could be a
supervisor outside a k8s cluster that wants to talk to the supervisor
inside the cluster.

So the definition of the rings would be a list of members and a tree
of zones (2). The parent-child relationship in the tree describes how
supervisor in one zone can communicate with supervisor on other
node. Taking the k8s cluster example from above, the supervisors
inside the cluster would be in a child zone of a zone that contains
supervisors outside the cluster. So supervisors in child zone can
easily talk to the supervisors in parent zone, but converse it's not
true.

A subset of supervisors in zone can act as gateways - these are the
supervisors that are exposed over the NAT boundary and these are the
supervisors to whom supervisors in parent zone can talk to in order to
get a message routed to some other supervisor that is not exposed.

So a zone could be described with an ID (UUID, just like a member), an
ID of a parent zone (possibly nil), a list of IDs of child zones and a
list of IDs of members which serve as gateways. (3)

When a supervisor starts, it has a nil zone ID. When it does not have
any peers to talk to, it will settle the zone ID before replying to
the first supervisor that talks to him, so the reply can possibly
contain a settled zone ID. By settling the zone ID I mean here
generating a new ID if it is nil.

In case of supervisor that just started and has a list of peers to
talk to to join the ring, it assumes that its zone ID will be the one
in the first reply it gets, unless the reply tells it not to do that -
in this case it will settle its zone ID.

The case when the supervisor receives a reply with a "do not take my
zone ID as yours" happens when the supervisor pinged the other one
behind a NAT. But how can the supervisor behind the NAT know that it
is about to send a reply to the supervisor on the other side of NAT
boundary? Adding a "To" field being an address and a port to the Ping
message could solve this issue, because the supervisor could compare
the internal address and port it already knows and the address from
the "To" field. If they are different then it very likely means that
the supervisor sending the Ping message should belong to a different
zone.

There may be a situation where a supervisor joins two separate rings,
so they should become one ring with one zone - at this moment, zone
merging should start. I suppose this could be done with a new kind of
rumor…

Also, the "To" field from the Ping message can serve as a way for the
supervisor to learn that it is really exposed outside the NAT, so the
supervisor becomes a gateway.

I think that the member description should be extended with
information about zone it belongs to and with a list of extra
addresses the member is available from.

The new member information and the new zones information could be used
to define the route between two supervisors in two different zones.

Probably the fastest way to do it would be to add another kind of
message (like ForwardedMessage) which could contain the payload (which
is either a SWIM or a gossip message), an ID of a member that this
message should go to and some TTL field. The ForwardedMessage message
could be the payload of a Wire message.

The recipient of the ForwardedMessage could check if the member ID is
the same as its own. If it's true, it could process it. Otherwise it
could compute a route and forward the message to the first supervisor
on the route. I suppose that if the supervisor doesn't have enough
information about how to forward the message, it could decrement the
TTL and send it to random supervisor in the same zone.

I haven't thought too much over the backward-compatibility, though. I
think that maybe it could be possible to detect that we are trying to
talk over NAT boundary to the supervisor that runs some old,
NAT-unaware version, and bail out or ignore or warn about it.

1. Can we safely assume that swim port and gossip port will always
have the same number? In butterfly it is possible for them to be
different, but this is not utilized by the supervisor - it uses the
same number for both. If they can be different in Real Life, it would
complicate the design, because a supervisor could not quickly discover
which port for gossip communication is being exposed after receiving a
Ping message over the exposed swim port.

2. A tree or a directed acyclic graph? Probably does not matter much
anyway.

3. The zone tree (or a graph) needs to be maintained somehow. Each
supervisor is responsible for maintaining its member information and
spreading information about the changes there. Who should be
responsible for maintaining the zone information? Yet another leader
chosen through election?

**Comment by asymmetric** [comment]
Could introducing the concept of zone be avoided if we assumed that each supervisor should be reachable from outside the cluster?

I think that the scaling downside is more theoretical than practical. What is the limit exactly? IIRC @krnowak you were afraid that in Kubernetes only around 2k ports could be opened for services, but that's [not the case](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport).

**Comment by krnowak** [comment]
The default limit is around 3k (30000-32767), but apparently you can change it with some flags. So I was wrong in this case.

The concept of zones was introduced only because I assumed that making each supervisor to be directly reachable over the NAT boundary is not always possible. Can there be a scenario not related to k8s where exposing every supervisor is not feasible?

**Comment by zeenix** [comment]
We (me, @LiliC and @krnowak) discussed this further today and we had a question that will help us decide on how to proceed further with this issue:

From what we understand, currently the supervisors provide their own IP themselves to their peers. If supervisors would use the IP they see on the packets (on the IP layer) instead of the one provided by the peer themselves, there will be no need for each supervisor to know its own IP address. This will allow us to enable communication between supervisors inside and outside a kubernetes cluster (or any supervisors separated by a NAT) through port forwarding and/or known NAT traversal protocols (for which there are libraries available).

So the question is if such a change to supervisor behavior be acceptable? We can keep things backwards compatible by only deprecating the IP information in the protocol so that supervisors keep providing their IP and existing supervisors will continue to use the IP advertised by the peer, but new supervisors will use the IP from the IP layer of the incoming packets.

**Comment by tashimi** [comment]
I think we should take a pause on this effort until the Supervisor work in flight lands to avoid a lot of re-work. Then we can take a look at how the Supervisor behavior has shifted and see how this would work.

**Comment by tashimi** [comment]
My high level notes on the Supervisor changes are here: https://ext.prodpad.com/ext/roadmap/d2938aed0d0ad1dd62669583e108357efd53b3a6

There are also some proposed changed to Butterfly that I documented here: https://github.com/habitat-sh/habitat/issues/4634

**Comment by tashimi** [comment]
@christophermaier has also been doing a bunch of discovery around how the Supervisors and Butterfly interact using data from @bixu and there appears to be an interesting disconnect where the gossiped data is more correct than the Supervisor, and regular patterns of irregularity in information occur.

**Comment by krnowak** [comment]
Note: When I'm talking about IP address I mean an IP:port pair.

TL;DR:

We will likely need to drop the idea of address discovery and add a possibility to specify public IP addresses for the supervisor and the services it manages if they are behind a NAT and we want to expose them. The public IP address would be used for communication with supervisors/services outside the NAT and internal IP address - for communication with supervisors/services behind the same NAT. I don't think we can employ NAT traversal algorithms here. I still need to figure out how to do it, so habitat-operator can use the feature.

> From what we understand, currently the supervisors provide their own IP themselves to their peers. If supervisors would use the IP they see on the packets (on the IP layer) instead of the one provided by the peer themselves, there will be no need for each supervisor to know its own IP address. This will allow us to enable communication between supervisors inside and outside a kubernetes cluster (or any supervisors separated by a NAT) through port forwarding and/or known NAT traversal protocols (for which there are libraries available).

That's not entirely correct I think. The supervisors send ring membership to each other. A part of this information is the IP address of a supervisor. And that information is actually taken from the IP layer (I assume that a result of [UdpSocket::recv_from](https://doc.rust-lang.org/std/net/struct.UdpSocket.html#method.recv_from) really comes from the IP layer, that is). So in the beginning, the supervisor doesn't really know it's IP address for SWIM/gossip communication (the IP address for services it manages is another thing, it is a bit of mess IMO) - the other supervisor it pings has that information first and it will be eventually propagated throughout the ring.

In code:

- the other supervisor receives a message together with an IP address of a sender: https://github.com/habitat-sh/habitat/blob/92c7de9a1bae5c46761320c0c60e1fca00d26a76/components/butterfly/src/server/inbound.rs#L61-L62
- if the message is Ping, we call a `process_ping` function where we pass the received message and IP address: https://github.com/habitat-sh/habitat/blob/92c7de9a1bae5c46761320c0c60e1fca00d26a76/components/butterfly/src/server/inbound.rs#L95
- we put the IP address into the member description and put it into the ring membership: https://github.com/habitat-sh/habitat/blob/92c7de9a1bae5c46761320c0c60e1fca00d26a76/components/butterfly/src/server/inbound.rs#L248-L260
- eventually this information will be propagated to other supervisors in the ring

There are two problems I think. First problem is that this works if we assume that the supervisor will always be available under this IP address and every other supervisor can talk to it using this IP address. These assumptions are false when the supervisor is behind a NAT - first assumption is false, because next time the supervisor behind a NAT talks outside, it can get a different IP address (well, same IP but different port number) assigned. Second assumption is false for symmetric NATs.

Second problem is that we are using UDP for SWIM communication and I'm not really sure how does it work with NAT traversal. If at all.

Putting it aside - there is another problem.

We don't form the spanning ring for the sake of having the spanning ring. We do it to advertise services that are managed by supervisors running behind the NAT to outside. So we need to expose the services too, not only supervisors. For supervisors, I had an idea of a discovery - a supervisor can learn about another IP address it is available under when the supervisor is being pinged, so it would know that it can serve as a gateway for other supervisors that want to send data across the NAT boundary. We can do it, because we are in full control of a supervisor code, but that idea won't fly for the services a supervisor manages - they are not our code. Instead, we need to expose the services explicitly (so they are accessible from outside the NAT), and then tell the supervisor about it, so it can generate a proper `Service` rumor for other supervisors in the ring to consume. That is - a rumor that can tell an application how to bind to the service, regardless of whether the application is running behind a NAT, just like the service, or outside. I think there is no such problem when, for example, an application is behind the NAT and the service it wants to bind to is outside. But I'm not sure - this is likely very application specific. All this in the end probably means that the `Service` rumor will need to contain multiple IP addresses, one per zone.

Anyway, if we need to be explicitly tell the supervisor about how the services it manages are exposed, then we can also explicitly tell it how the supervisor itself is exposed. That is - I could probably throw the discovery idea in the bin. And this needs to be done in a way that habitat-operator can use it inside k8s cluster.

I find it a bit of a hard nut to crack, because:

1. The services can have many different ways of connecting and we cannot cover them in some generic way I think.

But probably handling only the TCP and UDP ports should cover almost all of the cases.

2. We probably don't want to expose all the supervisors that are running behind the NAT. Nor we want to expose all the services running there I suppose.

The problem is that if we don't expose a service that is behind a NAT, other supervisors outside the NAT will receive information about the service anyway, through gossip, but the information will be useless - it will mention some unreachable, internal IP address. Maybe some filtering will need to be done.

Not covering supervisors here.

3. This probably will affect the way the service configuration is generated.

I mean a scenario where you have for example:

```
{{~#eachAlive bind.database.members as |member|}}
  database = "{{member.sys.ip}}:{{member.cfg.port}}"
  database-secure = "{{member.sys.ip}}:{{member.cfg.ssl-port}}"
{{~/eachAlive}}
```

Probably the `eachAlive` helper here should only return members that are accessible/exposed. Maybe this won't be necessary if we do filtering from previous point.

Likely there is more stuff/helpers that would need to handle the unexposed services/members.

**Comment by zeenix** [comment]
> First problem is that this works if we assume that the supervisor will always be available under this IP address and every other supervisor can talk to it using this IP address. These assumptions are false when the supervisor is behind a NAT - first assumption is false, because next time the supervisor behind a NAT talks outside, it can get a different IP address (well, same IP but different port number) assigned. 

I don't quite follow why being behind a NAT means you getting a different port? You can get a very specific port open with NAT-traversal and that's how it all works for games and realtime communications.

**Comment by krnowak** [comment]
In case of games and realtime communication it is likely that the computer behind a NAT establishes a connection to some publicly-accessible server and maintain the connection for as long as it is required. During the lifetime of the single connection, you'll get a single port, true, but I don't think you can assume that when the computer does many short-lived connections to servers outside. I'm not entirely sure here - zeromq is used in this case and I need to see what it does behind a curtain. OTOH, I heard/read somewhere that we want to drop zeromq.

**Comment by zeenix** [comment]
> In case of games and realtime communication it is likely that the computer behind a NAT establishes a connection to some publicly-accessible server and maintains the connection for as long as it is required. 

Connection? They use UDP.

**Comment by krnowak** [comment]
> Connection? They use UDP.

So yeah, I have read [rfc4787](https://tools.ietf.org/html/rfc4787). There's a bunch of requirements that a NAT has to fulfil for UDP over NAT to work. One of those was about the timeout after which the mapping between internal IP address to external IP address goes away. Games or realtime communication talk to the server constantly, so the timeout is constantly refreshed. I suppose I can say that SWIM communication also happens quite often, so the timeout won't happen easily.

Anyway, this point is kinda moot, I think, as games and realtime communications is the one that initiates the communication with the server. In case of supervisors, we want supervisor either behind a NAT or outside it to initiate it. For the latter to work, we need to expose the supervisor behind the NAT.
